### PR TITLE
rad-auth-keys CLI tool

### PR DESCRIPTION
**Pull request by Ryanmtate**
The `rad-auth-keys` CLI tool manages a proposed
`<git-repo>/.rad/keys` file for maintaining a keyring for namespace
access control.

This commit is part of a proof of concept for improving access control
to repositories through the use of validating git signed push
certificates evaluated in the `pre-receive` hook for the
`radicle-git-server`.

The intent of the `.rad/keys` directory is to live within the root of a
git repository, and to be updated with maintainers' GPG key fingerprints
to authorize a signed push request.

This means access control is managed by the maintainers accepting
changes to the `.rad/keys` directory.

Additionally, the structure of the `.rad/keys/` directory is made to be consistent
with other key management schemes, namely `patatt`.

# Basic usage:

In the root of your git project, run the following command to add a PGP key.

`gpg --armor --export <your-email> | rad-auth-keys add`

```
rad-auth-keys 0.1.0
Radicle authorized keys CLI tool for managing Radicle git repository authorized keys.

USAGE:
    rad-auth-keys [OPTIONS] <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -d, --dir <dir>              Optional, the path for parent directory of `.rad/` directory, defaults to
                                 `std::env::current_dir()?`
    -i, --id <id>                Optional, the signing key id (fingerprint) to add or remove to authorized keys list;
                                 required for `remove`
    -k, --key-type <key-type>    Optional, the signing key type used for authenticating a request, e.g. `openpgp`,
                                 `eip155:<chain_id>`, `ed25519`; defaults to `openpgp`
    -p, --path <path>            Optional, the path to the public key, otherwise will accept standard input for the
                                 public key
    -s, --source <source>        Optional, the source of the keyring to modify, e.g. `radkeys`, `radid`, `ens`; defaults
                                 to `radkeys`

SUBCOMMANDS:
    add       add a public key to the keyring source
    help      Prints this message or the help of the given subcommand(s)
    list      list the public keys in the keyring source
    remove    remove a public key from the keyring source
```

### CODE REVIEW

File: authorized-keys/src/rad_keys.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    }
+}
+
+impl RadKeys {
+    pub fn new(keys_dir: Option<PathBuf>) -> Result<Self, Error> {
+        let dir = match keys_dir {
+            Some(d) => d,
+            None => {
+                let mut dir = std::env::current_dir()?;
+                // Push to the default radical keys directory;
```
Comment (cloudhead) [inline authorized-keys/src/rad_keys.rs:51]:
```suggestion
                // Push to the default radicle keys directory;
```

File: authorized-keys/USAGE.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    -k, --key-type <key-type>    Optional, the signing key type used for authenticating a request, e.g. `openpgp`,
+                                 `eip155:<chain_id>`, `ed25519`; defaults to `openpgp`
+    -p, --path <path>            Optional, the path to the public key, otherwise will accept standard input for the
+                                 public key
+    -s, --source <source>        Optional, the source of the keyring to modify, e.g. `radkeys`, `radid`, `ens`; defaults
+                                 to `radkeys`
+
+SUBCOMMANDS:
+    add       add a public key to the keyring source
+    help      Prints this message or the help of the given subcommand(s)
```
Comment (cloudhead) [inline authorized-keys/USAGE.md:28]:
```suggestion
    help      print this message or the help of the given subcommand(s)
```

File: authorized-keys/src/rad_keys.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+use crate::{
+    error::{Error, TIMEOUT_STDIN_WARNING},
+    Action, KeyType, Options, RadKeyring,
+};
+
+// Default directory for storing keys;
+const DEFAULT_RAD_KEYS_DIRECTORY: &str = ".rad/keys/";
+
+/// git branch name
+pub type Branch = String;
```
[the remark this answers is not in the document]
Comment (Ryanmtate) [reply authorized-keys/src/rad_keys.rs]: commit db46635 removes.

**Comment by Ryanmtate** [comment]
> @cloudhead Is this ready to merge? Any additional feedback?

I might re-work this tonight to tree the `rad auth keys` and separate subcommands to make it easier to merge with the eventual `rad` main CLI.

**Comment by cloudhead** [comment]
Thanks!

Merged in 59feb7d8bfd913fd2389a689f64f834c5b71ccd7

**Pull request closed without merging**